src/hotspot/share/cds/heapShared.cpp

```diff
@@ -781,6 +781,137 @@ void HeapShared::copy_objects() {
   }
 }
 
+// Mark the classes of all objects that are scanned HeapShared::find_all_aot_initialized_classes()
+class HeapShared::AOTInitializedClassScanner : public BasicOopIterateClosure {
+  bool _made_progress;
+
+  template <class T> void check(T *p) {
+    oop obj = HeapAccess<>::oop_load(p);
+    if (!java_lang_Class::is_instance(obj)) {
+      // Don't scan the mirrors, as we may see an orig_mirror while scanning
+      // the object graph, .... TODO more info
+      _made_progress |= HeapShared::scan_for_aot_initialized_classes(obj);
+    }
+  }
+
+public:
+  AOTInitializedClassScanner() : _made_progress(false) {}
+  void do_oop(narrowOop *p) { check(p); }
+  void do_oop(      oop *p) { check(p); }
+  bool made_progress() { return _made_progress; }
+};
+
+static bool mark_for_aot_initialization(InstanceKlass* buffered_ik) {
+  if (buffered_ik->name()->equals("java/lang/String")) {
+    return false;
+  }
+
+  assert(ArchiveBuilder::current()->is_in_buffer_space(buffered_ik), "sanity");
+
+  bool made_progress = false;
+  if (buffered_ik->is_initialized() && !buffered_ik->has_aot_initialized_mirror()) {
+    if (log_is_enabled(Info, cds, init)) {
+      ResourceMark rm;
+      log_info(cds, init)("Mark class for aot-init: %s", buffered_ik->external_name());
+    }
+
+    buffered_ik->set_has_aot_initialized_mirror();
+    made_progress = true;
+
+    InstanceKlass* super = buffered_ik->java_super();
+    if (super != nullptr) {
+      mark_for_aot_initialization(super);
+    }
+
+    Array<InstanceKlass*>* interfaces = buffered_ik->local_interfaces();
+    for (int i = 0; i < interfaces->length(); i++) {
+      mark_for_aot_initialization(interfaces->at(i));
+    }
```

Comment:
This may initialize a larger set of interfaces than required to support the initialization of `buffered_ik`.  

JVMS 5.5 "Initialization" only requires the interfaces that have default methods (more actually, at least one `non-abstract non-static`) to be initialized as part of `buffered_ik`'s initialization.

Are we looking for a minimal set of classes to AOT initialize or a maximal set?  We should be clear on which one we are picking and preferably add a comment to the code on why we picked it

Reply:
I've updated the code in the latest patch to make sure that uninitialized interfaces are not marked as aot-initialized.

https://github.com/openjdk/jdk/pull/21143/files#diff-993b1b2c5ad46fbd044f92e2e82a1dea88493fede873095602492f19ebeacb5cR919-R927

Note that this function doesn't change the initialization state of any classes (it can't execute Java code as we are in a safepoint).